### PR TITLE
Implement ability to limit module documentation building

### PR DESCRIPTION
**Pull request by azaghal**
##### SUMMARY

The purpose of this pull request is to introduce option for limiting which module documentation gets built. Implementation achieves this via `MODULES` environment variable that can get passed-in to `make webdocs`.

The reason for this change is to allow both module developers and documentation contributors to more quickly see and test their changes. Building entire documentation for all modules can take quite a while, and it deters from fixing small syntax, typos and similar errors.

##### ISSUE TYPE
 - Docs Pull Request

##### COMPONENT NAME
Documentation build system

##### ANSIBLE VERSION

```
ansible 2.4.0 (doc_module_selection 7b6dd1f4d0) last updated 2017/05/13 13:43:50 (GMT +200)
  config file = 
  configured module search path = [u'REDACTED', u'/usr/share/ansible/plugins/modules']
  ansible python module location = REDACTED
  executable location = REDACTED
  python version = 2.7.12 (default, Dec 13 2016, 14:29:11) [GCC 4.9.3]
```

##### ADDITIONAL INFORMATION

In order to test the change, you can do the following:

```
cd docs/docsite/

# Build documentation for apt module.
MODULES=apt make webdocs

# Build documentation for a couple of modules.
MODULES=apt,setup,yum make webdocs

# Don't build documentation for modules.
MODULES=none make webdocs

# Build documentation for all modules.
MODULES=all make webdocs
make webdocs
```

### CODE REVIEW

**Comment by azaghal** [comment]
Ok, next time I should run the sanity test myself :P

**Review by dharmabumstead — APPROVED** [review]
Fantastic - thanks @azaghal!

**Comment by dharmabumstead** [comment]
Looks good to me. Someone from the core team please review and merge if no issues.

**Comment by bcoca** [comment]
making it a 'limit' but also forcing to specify seem counter indicative, I would either rename to 'list' or would default to None/absent and filter against list if it is passed in.

Also i would make into list on args parsing, not in the function that is rendering.

other than that minutia, looks good

File: docs/bin/plugin_formatter.py
Comment (alikins) [inline docs/bin/plugin_formatter.py]: I think I would make limit_to_modules arg a list here, and do the comma split on the cli arg before calling list_modules. 

(currently it looks this would do odd things if attempting to build module docs for a module with a name like 'ec2_vps_cwt_jkg_none_facts' or 'firewalld' or 'easy_install'?  Haven't tested though so that may be covered...)

Comment (azaghal) [reply docs/bin/plugin_formatter.py]: `limit_to_modules` is currently supposed to be comma-separated string, so modules with mentioned names should not be a problem.

Comment (azaghal) [reply docs/bin/plugin_formatter.py]: I'll see what I can do to switch to lists and conditionals. One problem is my `Makefile` expertise is virtually non-existent (and don't have much time/will to learn it atm :)

Comment (azaghal) [reply docs/bin/plugin_formatter.py]: Hm... Another thing is how to specify "no modules" at all. I.e. without using the "all" or "none" keywords? Or would it be ok to preserve a variant where value of `all` or `none` means that?

Comment (azaghal) [reply docs/bin/plugin_formatter.py]: Sheesh... So, before I get any deeper into this, the `limit_to_modules` parameter received after option parser is invoked should end-up producing one of the following values:

- `None` - no limitation, default if option is not specified.
- `[]` - don't build any modules. This should be set if you specify special keyword - `--limit-to-modules none`.
- `['a', 'b', 'c']` - build only specified module. This should be set if you specify `--limit-to-modules a,b,c`.

Three ways that I can see to go around this (with one simply not working):

1. Massage option value after call to `(options, args) = p.parse_args()`.

2. [Add a new type to option parser](https://docs.python.org/2/library/optparse.html#adding-new-types).

3. Use callback, but this won't work since we need to set type, and there is no `list` type (plus, we need `list + None` type in this particular case I guess).

Preferences? Or am I just overcomplicating this?

**Comment by azaghal** [comment]
Ok, updated the pull request based on comments from review (gave-up on complicated solution :).

File: docs/bin/plugin_formatter.py
Comment (bcoca) [inline docs/bin/plugin_formatter.py]: making None the default saves you some code

Comment (azaghal) [reply docs/bin/plugin_formatter.py]: Ah, right... Was playing around with setting `type`, and that one did not like None. Will fix.

Comment (azaghal) [reply docs/bin/plugin_formatter.py]: Actually... I don't think it does, but I will fix it.

Comment (azaghal) [reply docs/bin/plugin_formatter.py]: Pushed again.

File: docs/bin/plugin_formatter.py
Comment (alikins) [inline docs/bin/plugin_formatter.py:457]: dont reassign options.limit_to_modules, just created a new limit_to_modules var

``` python
if options.limit_to_modules is not None:
    limit_to_modules = [s.lower() for s in options.limit_to_modules.split(",")]

mod_info, categories, aliases = list_modules(options.module_dir, limit_to_modules=limit_to_modules)

```

**Comment by gundalow** [comment]
so @nitzmahone https://github.com/ansible/ansible/pull/25313 gives the ability to generate a single non-module html page. Combined with this we can generate any page.

This would save lots of people lots of time, so +1000 from me

File: docs/bin/plugin_formatter.py
Comment (abadger) [reply docs/bin/plugin_formatter.py:457]: I wouldn't block on this but stylewise, I do agree with alikins that it's better to do all modifications of things in options inside of the argument parsing functions and then set other variables if the code is outside of it.   (In my own code I tend to have a parse_options function that would combine generate_parser, validate_options, and also do transformations on the arguments like this.  That function would return the options data structure at the end and then I'd try to treat it as immutable thereafter).

**Comment by abadger** [comment]
+1 for this.  Note that running html docs build with a subset of modules can lead to false positive errors (due to modules being able to link to other modules and those other modules perhaps not being present).  So you do have to be careful if you add this to some sort of automated testing.

**Comment by jhawkesworth** [comment]
Likely we don't need https://github.com/ansible/ansible/pull/25791 if this gets merged, so cross referencing it.
Would perhaps be nice if this wasn't limited just to modules and you could use it to build a page in the documentation too.

**Comment by azaghal** [comment]
Given that #25313 got merged, not sure this pull request is relevant? I actually thought I had closed it already :)

**Comment by gundalow** [comment]
What's the plan for this & https://github.com/ansible/ansible/pull/25791
Looks like we are very close to amazingness, just need to get the ability to build a single module done and then update the docs.

**Comment by azaghal** [comment]
Well, at least I have no idea what the plans are. I took a bit different approach compared to #25313, more centered around modules. For what it's worth, my patch is faster on initial build (since processing of most modules is skipped). Otherwise, the other approach covers everything (minus the docs :)

**Comment by jhawkesworth** [comment]
I updated #25791 with docs and shellcheck fixes now.
I just wanted an easy-to-remember command I could run to build a bit of the docs as a full build of the docs takes longer than my daily train journey.  No objection to making ansible-makepage.sh faster though, if it gets accepted.

File: docs/docsite/README.md
Comment (mattclay) [inline docs/docsite/README.md:21]: I recommend passing vars as args in all new `make` examples. Example: `make webdocs MODULES=none`

Also, use `a non-existent module name` instead of `non-existing module name`.

Not a blocker, just a suggestion to improve readability. If you do change it, update all new occurrences.

**Comment by gundalow** [comment]
We've agreed to go with this and close https://github.com/ansible/ansible/pull/25791

**Review by dharmabumstead — APPROVED** [review]
Docs OK with some minor edits. Can't wait to implement this. Thanks @azaghal!

**Comment by azaghal** [comment]
Sorry for late reply (as usual...), but going through my mail backlog :)

Should I continue in the same direction, probably just double-checking on reviews? I am happy to invest time as long as it's not a dead (or super-prolonged) end :)

**Comment by dharmabumstead** [comment]
Thanks very much @azaghal!

**Comment by azaghal** [comment]
Oh, no changes needed in the end? Great, I hope it will help people :)

**Pull request merged**
